File: contrib/android/app/src/main/java/org/pcoin/miner/Fmt.kt
package org.pcoin.miner

import java.util.Locale

/**
 * Display formatting.
 *
 * Deliberately dumb: it prints what the node reported and nothing else. There
 * are no earnings projections, no fiat conversions and no extrapolated "coins
 * per day" anywhere in this app, because none of those numbers would be real.
 */
object Fmt {

    /** Hash rate. RandomX on a phone lands in the tens-to-hundreds of H/s. */
    fun hashrate(hs: Double): String = when {
        hs.isNaN() || hs < 0 -> "--"
        hs < 1_000 -> String.format(Locale.US, "%.1f H/s", hs)
        hs < 1_000_000 -> String.format(Locale.US, "%.2f kH/s", hs / 1_000.0)
        else -> String.format(Locale.US, "%.2f MH/s", hs / 1_000_000.0)
    }

    /** Coin amounts, as reported by the wallet. -1 means "not known yet". */
    fun coins(v: Double): String =
        if (v.isNaN() || v < 0) "--" else String.format(Locale.US, "%.8f PCN", v)

    /**
     * Coin amounts held in satoshi. Everything on the forwarding path works in
     * integers -- a transaction is built, verified and recorded in satoshi, and
     * only the display converts -- because a fee assertion done in floating
     * point is an assertion that can be off by a rounding error.
     */
    fun coinsSat(sat: Long): String =
        if (sat < 0) "--" else String.format(Locale.US, "%d.%08d PCN", sat / 100_000_000L, sat % 100_000_000L)

    /**
     * A rough wall-clock duration, for "about 10 h from now" style estimates.
     * Deliberately coarse: the underlying number comes from observed block
     * spacing on a chain whose difficulty is currently pinned, so minutes of
     * precision would be false confidence.
     */
    fun roughDuration(ms: Long): String = when {
        ms < 0 -> "--"
        ms < 90 * 60_000L -> "${maxOf(1L, ms / 60_000L)} min"
        ms < 48 * 3_600_000L -> "${Math.round(ms / 3_600_000.0)} h"
        else -> "${Math.round(ms / 86_400_000.0)} days"
    }

    fun temp(c: Float): String =
        if (c.isNaN()) "--" else String.format(Locale.US, "%.1f°C", c)

    fun count(v: Long): String = if (v < 0) "--" else v.toString()

    fun count(v: Int): String = if (v < 0) "--" else v.toString()

    /** Chain height, with sync progress when the node is still catching up. */
    fun height(height: Long, headers: Long, progress: Double): String {
        if (height < 0) return "--"
        val base = height.toString()
        val syncing = headers > height || (progress in 0.0..0.9999)
        if (!syncing) return base
        val pct = if (progress in 0.0..1.0) {
            String.format(Locale.US, " (syncing %.1f%%)", progress * 100.0)
        } else if (headers > height) {
            " (syncing, $headers headers)"
        } else {
            " (syncing)"
        }
        return base + pct
    }

    /**
     * "50%  (4 of 8 cores)" -- deliberately the same wording the Windows tray
     * app uses (contrib/windows-tray/PCoinTray.cs), so the two clients describe
     * the same setting the same way.
     */
    fun performance(percent: Int, threads: Int, cores: Int): String =
        String.format(Locale.US, "%d%%  (%d of %d cores)", percent, threads, cores)
}

File: contrib/ios/PCoinKit/Sources/PCoinKit/Fmt.swift
import Foundation

/// Display formatting.
///
/// Deliberately dumb, exactly as the Android `Fmt.kt` is: it prints what was
/// read and nothing else. No earnings projections, no fiat conversions, no
/// extrapolated "coins per day" anywhere in this app, because none of those
/// numbers would be real.
///
/// Note what is missing. There is no hashrate, no temperature, no thread count
/// and no performance line: this is the wallet, and there is no miner on iOS.
/// Apple prohibits on-device mining and one must never be added.
public enum Fmt {

    /// Coin amounts held in satoshi.
    ///
    /// Everything works in integers -- a transaction is built, verified and
    /// recorded in satoshi and only the display converts -- because a fee
    /// assertion done in floating point is an assertion that can be off by a
    /// rounding error.
    ///
    /// A negative value means "not known yet" and prints an em dash. It is never
    /// printed as 0.00000000: a balance nobody has managed to read is not a
    /// balance of nothing.
    public static func coinsSat(_ sat: Int64) -> String {
        sat < 0 ? unknown : Amounts.toPlainString(sat) + " PCN"
    }

    /// The em dash the whole app uses for "not known".
    public static let unknown = "\u{2014}"

    public static func count(_ v: Int) -> String { v < 0 ? unknown : String(v) }

    /// A rough wall-clock duration, for "updated 5 min ago" style lines.
    ///
    /// Deliberately coarse. It is used for how long ago something was READ,
    /// never for how long something will take: this chain block spacing is
    /// noisy and routinely far from its 600 s target in both directions, so an
    /// ETA rendered from a hardcoded spacing is a guess wearing a number
    /// clothes.
    public static func roughDuration(seconds: TimeInterval) -> String {
        let ms = seconds * 1000
        switch ms {
        case ..<0: return unknown
        case ..<(90 * 60_000): return "\(max(1, Int(ms / 60_000))) min"
        case ..<(48 * 3_600_000): return "\(Int((ms / 3_600_000).rounded())) h"
        default: return "\(Int((ms / 86_400_000).rounded())) days"
        }
    }

    /// "Block 7342 - 51 peers", or the not-connected line.
    ///
    /// Both come from the explorer index block. On Android this describes the
    /// node inside the app; here it describes the node the explorer is reading,
    /// which is a different claim and the wording on screen says so.
    public static func chainLine(height: Int?, behind: Int?) -> String? {
        guard let h = height else { return nil }
        if let b = behind, b > 0 { return "Block \(h) - index \(b) behind" }
        return "Block \(h)"
    }

    /// Confirmations, as the history screen states them.
    ///
    /// nil in means nil out. A transaction whose confirmation count could not be
    /// read is not a transaction with zero confirmations -- that conversion is
    /// the one that authorises spending the same coins twice.
    public static func confirmations(_ n: Int?) -> String {
        guard let n = n else { return unknown }
        if n <= 0 { return "Waiting to be included in a block" }
        return n == 1 ? "1 confirmation" : "\(n) confirmations"
    }

    /// A shortened address for a tight row: first 12 and last 8, joined by an
    /// ellipsis. Never used where the destination is being CONFIRMED -- the
    /// review screen shows the whole thing, because a lookalike address is the
    /// attack and the eye is the only defence against it.
    public static func shortAddress(_ a: String) -> String {
        guard a.count > 24 else { return a }
        return a.prefix(12) + "\u{2026}" + a.suffix(8)
    }
}
